File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/flow/FrameState.java
/*******************************************************************************
 * Copyright (c) 2015-2018 Skymind, Inc.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

package org.nd4j.autodiff.samediff.flow;

import lombok.Data;
import lombok.NonNull;

/**
 * This class is a holder for state of loops imported from TensorFlow, via frame_name
 *
 * @author raver119@gmail.com
 */
@Data
public class FrameState {
    private String name;
    private long iterations = 0;
    private boolean rewindPlanned = false;
    private int rewindPosition = -1;

    private int numberOfEntries = 0;
    private int numberOfExits = 0;
    private boolean active = false;
    private int numberOfCycles;


    public FrameState(@NonNull String frame_name) {
        this.name = frame_name;
    }

    /**
     * This method returns number of cycles for this frame
     * @return
     */
    public int getNumberOfCycles() {
        return numberOfCycles;
    }

    /**
     * This method increments number of cycles by 1
     */
    public void incrementNumberOfCycles() {
        numberOfCycles++;
    }
}

File: libnd4j/include/graph/impl/FrameState.cpp
/*******************************************************************************
 * Copyright (c) 2015-2018 Skymind, Inc.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// Created by raver119 on 06.02.2018.
//

#include <graph/FrameState.h>


namespace nd4j {
    namespace graph {
        FrameState::FrameState(Nd4jLong id) {
            this->_id = id;
        }

        int FrameState::getNumberOfCycles() {
            return _numberOfCycles;
        }

        void FrameState::incrementNumberOfCycles() {
            ++_numberOfCycles;
        }

        bool FrameState::wasActivated() {
            return _activated;
        }

        void FrameState::markActivated(bool reallyActivated) {
            _activated = reallyActivated;
        }

        std::string &FrameState::getFrameName() {
            return _name;
        }

        bool FrameState::isRewindPlanned() {
            return _rewindPlanned;
        }

        int FrameState::getRewindPosition() {
            return _rewindPosition;
        }

        void FrameState::setRewindPosition(int pos) {
            _rewindPosition = pos;
        }

        void FrameState::setRewindPositionOnce(int pos) {
            if (_rewindPosition < 0)
                _rewindPosition = pos;
        }

        void FrameState::planRewind(bool reallyPlanning) {
            _rewindPlanned = reallyPlanning;
        }
    }
}